File: flux/apps/games/futboss-ai/packages/shared/src/types/team.ts
// FutBoss AI - Team Types
// Author: Bruno Lucena (bruno@lucena.cloud)

export type Formation = '4-4-2' | '4-3-3' | '3-5-2' | '4-5-1' | '3-4-3' | '5-3-2';

export interface Team {
  id: string;
  name: string;
  ownerId: string;
  formation: Formation;
  playerIds: string[];
  wins: number;
  draws: number;
  losses: number;
  goalsFor: number;
  goalsAgainst: number;
  createdAt: Date;
}

export interface TeamCreate {
  name: string;
  formation?: Formation;
}

export interface TeamResponse extends Team {
  points: number;
  goalDifference: number;
}

export function calculatePoints(team: Team): number {
  return team.wins * 3 + team.draws;
}

export function calculateGoalDifference(team: Team): number {
  return team.goalsFor - team.goalsAgainst;
}

export function calculateMatchesPlayed(team: Team): number {
  return team.wins + team.draws + team.losses;
}

export function calculateWinRate(team: Team): number {
  const matches = calculateMatchesPlayed(team);
  if (matches === 0) return 0;
  return (team.wins / matches) * 100;
}

File: flux/apps/games/futboss-ai/apps/api/app/models/team.py
# FutBoss AI - Team Model
# Author: Bruno Lucena (bruno@lucena.cloud)

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class Formation(str, Enum):
    F442 = "4-4-2"
    F433 = "4-3-3"
    F352 = "3-5-2"
    F451 = "4-5-1"
    F343 = "3-4-3"
    F532 = "5-3-2"


class TeamCreate(BaseModel):
    name: str = Field(..., min_length=3, max_length=50)
    formation: Formation = Formation.F442


class Team(BaseModel):
    id: Optional[str] = Field(default=None, alias="_id")
    name: str
    owner_id: str
    formation: Formation = Formation.F442
    player_ids: List[str] = Field(default_factory=list)
    wins: int = 0
    draws: int = 0
    losses: int = 0
    goals_for: int = 0
    goals_against: int = 0
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True

    def points(self) -> int:
        return (self.wins * 3) + self.draws

    def matches_played(self) -> int:
        return self.wins + self.draws + self.losses

    def goal_difference(self) -> int:
        return self.goals_for - self.goals_against


class TeamResponse(BaseModel):
    id: str
    name: str
    owner_id: str
    formation: Formation
    player_ids: List[str]
    wins: int
    draws: int
    losses: int
    points: int
    goal_difference: int

    class Config:
        from_attributes = True

